flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java

```diff
@@ -297,9 +318,279 @@ private static Object handleLogicalTypeSchema(Schema fieldSchema, Object value)
+                        String.format(
+                                "Invalid Timestamp '%s' Provided."
+                                        + "\nShould be a long value indicating microseconds since Epoch (1970-01-01 00:00:00.000000+00:00) "
+                                        + "between %s and %s",
+                                timestamp, minTs, maxTs));
+            }
+        }
+
+        // BigQuery inputs timestamp as microseconds since EPOCH,
+        // So if we have TIMESTAMP in micros - we convert as it is.
+        // If the TIMESTAMP is in millis - we convert to Micros and then add.
+        @VisibleForTesting
+        static Long convertTimestamp(Object value, boolean micros, String type) {
+            long timestamp;
+            if (value instanceof ReadableInstant) {
+                timestamp = TimeUnit.MILLISECONDS.toMicros(((ReadableInstant) value).getMillis());
+            } else {
+                Preconditions.checkArgument(
+                        value instanceof Long,
+                        String.format(
+                                "Expecting a value as Long type %s. Instead %s was obtained",
+                                type, value.getClass()));
+                timestamp = (micros ? (Long) value : TimeUnit.MILLISECONDS.toMicros((Long) value));
+            }
+            return validateTimestamp(timestamp);
+        }
+
+        private static Integer validateDate(Integer date) {
+            if (date > 2932896 || date < -719162) {
+                throw new IllegalArgumentException(
+                        String.format(
+                                "Invalid Date '%s' Provided."
+                                        + "\nShould be a Integer value indicating days since Epoch (1970-01-01 00:00:00) "
+                                        + "between %s and %s",
+                                LocalDate.ofEpochDay(date), "0001-01-01", "9999-12-31"));
+            }
+            return date;
+        }
+
+        @VisibleForTesting
+        static Integer convertDate(Object value) {
+            // The value is the number of days since the Unix epoch (1970-01-01).
+            // The valid range is `-719162` (0001-01-01) to `2932896` (9999-12-31).
+            int date;
+            if (value instanceof ReadableInstant) {
+                date = Days.daysBetween(Instant.EPOCH, (ReadableInstant) value).getDays();
+            } else {
+                Preconditions.checkArgument(
+                        value instanceof Integer, "Expecting a value as Integer type (days).");
+                date = (Integer) value;
+            }
+            return validateDate(date);
+        }
+
+        @VisibleForTesting
+        static String convertDateTime(Object value, boolean micros) {
+            if (value instanceof String) {
+                return (String) value;
+            }
+            if (value instanceof Utf8) {
+                return ((Utf8) value).toString();
+            }
```

Comment:
These values do not need to be validated?

Reply:
They do, while implementing had created bugs for the same (b/329265127 and b/329266379). Since BQ allows time and datetime literal in various formats, it is a non trivial implementation.  Please let me know in case of a existing implementation for the same.